inseq/attr/feat/ops/rollout.py

```diff
@@ -0,0 +1,139 @@
+            the scores aggregated using rollout until the topmost provided layer (e.g. for ``layers=[1,2,4]`` the
+            rollout is done skipping layer 3, and only rolled out scores at layer 4 are returned). If encoder-decoder
+            rollout is performed, a tuple of tensors ``(source_scores, target_scores)``.
+    """
+    squeeze_batch_dim = False
+    remove_padding = False
+    if isinstance(scores, tuple):
+        if dim < 0:
+            dim = scores[0].ndim + dim
+        if scores[0].ndim < 4:
+            scores = tuple(t.unsqueeze(0) for t in scores)
+            squeeze_batch_dim = True
+            dim += 1
+        if dim != 1:
+            swap_dim = dim if dim < 2 else dim + 1
+            scores = tuple(s[:, None, ...].transpose(swap_dim, 1).squeeze(swap_dim) for s in scores)
+        source_scores, cross_scores, target_scores = scores
+
+        # Get rolled out scores of encoder last layer with respect to source input
+        final_source_scores = _rollout_single(source_scores.mT)[:, -1, ...].mT
+
+        source_rollout_scores, target_rollout_scores = _rollout_joint(final_source_scores, cross_scores, target_scores)
+        source_rollout_scores = source_rollout_scores[:, -1, ...].unsqueeze(1)
+        target_rollout_scores = target_rollout_scores[:, -1, ...].unsqueeze(1)
+        if dim != 1:
+            source_rollout_scores = source_rollout_scores.transpose(1, dim)
+            target_rollout_scores = target_rollout_scores.transpose(1, dim)
+        source_rollout_scores = source_rollout_scores.squeeze(dim)
+        target_rollout_scores = target_rollout_scores.squeeze(dim)
+        if squeeze_batch_dim:
+            source_rollout_scores = source_rollout_scores.squeeze(0)
+            target_rollout_scores = target_rollout_scores.squeeze(0)
+        return source_rollout_scores, target_rollout_scores
+    else:
+        # Convert rollout dim to positive index to account for new dim insertions.
+        if dim < 0:
+            dim = scores.ndim + dim
+        # Add batch dimension if not present. Assumed shape (batch_size, ...) with num_layers at position dim and at
+        # least two dimensions representing scores that will be rolled out.
+        if scores.ndim < 4:
+            scores = scores[None, ...]
+            squeeze_batch_dim = True
+            dim += 1
+        if dim != 1:
+            scores = scores[:, None, ...]
+            swap_dim = dim if dim < 2 else dim + 1
+            scores = scores.transpose(swap_dim, 1).squeeze(swap_dim)
+        scores[scores.isnan()] = 0.0
+        # If the matrix is not square (e.g. generating from a prefix), prepend zeros to make it square.
+        if scores.size(-1) < scores.size(-2):
+            pad_size = scores.size(-2) - scores.size(-1)
+            scores = torch.cat([torch.zeros(*tuple(scores.shape[:-1]), pad_size), scores], dim=-1)
+            remove_padding = True
```

Comment:
This is wrong since the absence of scores for the prefix prevents the application of rollout. This is a problem for decoder-only models generating from a prefix, and in general any attribution where the starting attribute step is not the actual BOS.

A possible fix would involve the following steps:
- Enable decoder-only unconstrained generation with empty inputs (requires fix related to #190)
- Add a new `model.attribute` parameter (e.g. `attribute_full_generation: bool = False`) that when set to `True` attribute the full generation, regardless of the specified target prefix. This enables treating the target prefix as part of the generated sequence, both for decoder-only (common case) and encoder-decoders (overriding `attr_pos_start` with a warning, or when continuing generating from a target prefix), ultimately getting all scores needed for rollout as outputs
- Raise an exception in rollout if the matrices used are not exactly `S x S`, `S x T` and `T x T` (encoder-decoder case) or `T x T` (decoder-only case), where `S` is source length and `T` is target length, mentioning that `attribute_full_generation` should be set in that case to enable rollout